pkg/services/authn/clients/jwt.go

```diff
@@ -0,0 +1,208 @@
+	jwtToken = strings.TrimPrefix(jwtToken, "Bearer ")
+
+	claims, err := s.jwtService.Verify(ctx, jwtToken)
+	if err != nil {
+		s.log.Debug("Failed to verify JWT", "error", err)
+		return nil, ErrJWTInvalid.Errorf("failed to verify JWT")
+	}
+
+	sub, _ := claims["sub"].(string)
+	if sub == "" {
+		s.log.Warn("Got a JWT without the mandatory 'sub' claim", "error", err)
+		return nil, ErrJWTMissingClaim.Errorf("missing mandatory 'sub' claim in JWT")
+	}
+
+	id := &authn.Identity{
+		AuthModule: login.JWTModule,
+		AuthID:     sub,
+		OrgRoles:   map[int64]org.RoleType{},
+		ClientParams: authn.ClientParams{
+			SyncUser:            true,
+			SyncTeamMembers:     true,
+			AllowSignUp:         false,
+			EnableDisabledUsers: false,
+		}}
+
+	if key := s.cfg.JWTAuthUsernameClaim; key != "" {
+		id.Login, _ = claims[key].(string)
+		id.ClientParams.LookUpParams.Login = &id.Login
+	}
+	if key := s.cfg.JWTAuthEmailClaim; key != "" {
+		id.Email, _ = claims[key].(string)
+		id.ClientParams.LookUpParams.Email = &id.Email
+	}
+
+	if name, _ := claims["name"].(string); name != "" {
+		id.Name = name
+	}
+
+	role, grafanaAdmin := s.extractRoleAndAdmin(claims)
+	if s.cfg.JWTAuthRoleAttributeStrict && !role.IsValid() {
+		s.log.Warn("extracted Role is invalid", "role", role, "auth_id", id.AuthID)
+		return nil, ErrJWTInvalidRole.Errorf("invalid role claim in JWT: %s", role)
+	}
+
+	if role.IsValid() {
+		var orgID int64
+		// FIXME (jguer): GetIDForNewUser already has the auto assign information
+		// just neeeds the org role. Find a meaningful way to pass this default
+		// role to it (that doesn't involve id.OrgRoles[0] = role)
+		if s.cfg.AutoAssignOrg && s.cfg.AutoAssignOrgId > 0 {
+			orgID = int64(s.cfg.AutoAssignOrgId)
+			s.log.Debug("The user has a role assignment and organization membership is auto-assigned",
+				"role", role, "orgId", orgID)
+		} else {
+			orgID = int64(1)
+			s.log.Debug("The user has a role assignment and organization membership is not auto-assigned",
+				"role", role, "orgId", orgID)
+		}
+
+		id.OrgRoles[orgID] = role
+		if s.cfg.JWTAuthAllowAssignGrafanaAdmin {
+			id.IsGrafanaAdmin = &grafanaAdmin
+		}
+	}
+
+	if id.Login == "" || id.Email == "" {
+		s.log.Debug("Failed to get an authentication claim from JWT",
+			"login", id.Login, "email", id.Email)
+		return nil, ErrJWTMissingClaim.Errorf("missing login or email claim in JWT")
+	}
+
+	if s.cfg.JWTAuthAutoSignUp {
+		id.ClientParams.AllowSignUp = true
+	}
+
+	return id, nil
+}
+
+func (s *JWT) Test(ctx context.Context, r *authn.Request) bool {
+	if !s.cfg.JWTAuthEnabled || s.cfg.JWTAuthHeaderName == "" {
+		return false
+	}
+
+	jwtToken := r.HTTPRequest.Header.Get(s.cfg.JWTAuthHeaderName)
+	if jwtToken == "" && s.cfg.JWTAuthURLLogin {
+		jwtToken = r.HTTPRequest.URL.Query().Get("auth_token")
+	}
+
+	if jwtToken == "" {
+		return false
+	}
+
+	// Strip the 'Bearer' prefix if it exists.
+	jwtToken = strings.TrimPrefix(jwtToken, "Bearer ")
```

Comment:
nit: These are almost identical to https://github.com/grafana/grafana/pull/61157/files#diff-da754e78d43b1b454b09a7ee9163959b631d7622cbe7a9f495fc0acca0032ddbR45-R51. So the `jwtToken` extraction logic could be extracted to a function.